proposals/gc/MVP.md

```diff
@@ -0,0 +1,312 @@
+  - `optref $t <: anyref`
+  - `nullref <: optref $t`
+
+* Any concrete reference type is a subtype of the respective nullable reference type (and thereby of `anyref`)
+  - `ref $t <: optref $t`
+  - Note: concrete reference types are *not* supertypes of `nullref`, i.e., not nullable
+
+* Any function reference type is a subtype of `anyfunc`
+  - `ref $t <: anyfunc`
+     - iff `$t = <functype>`
+
+* Any optional reference type (and thereby respective concrete reference type) is a subtype of `eqref` if its not a function
+  - `optref $t <: eqref`
+     - if `$t = <structtype>` or `$t = <arraytype>`
+     - or `$t = type rt` and `rt <: eqref`
+  - TODO: provide a way to make data types non-eq, especially immutable ones
+
+* Concrete and optional reference types are covariant
+  - `ref $t1 <: ref $t2`
+     - iff `$t1 <: $t2`
+  - `optref $t1 <: optref $t2`
+     - iff `$t1 <: $t2`
+
+* Structure types support width and depth subtyping
+  - `struct <fieldtype1>* <fieldtype1'>* <: struct <fieldtype2>*`
+    - iff `(<fieldtype1> <: <fieldtype2>)*`
+
+* Array types support depth subtyping
+  - `array <fieldtype1> <: array <fieldtype2>`
+    - iff `<fieldtype1> <: <fieldtype2>`
+
+* Field types are covariant if they are immutable, invariant otherwise
+  - `const <valtype1> <: const <valtype2>`
+    - iff `<valtype1> <: <valtype2>`
+  - `var <valtype> <: var <valtype>`
+  - Note: mutable fields are *not* subtypes of immutable ones, so `const` really means constant, not read-only
+
+
+#### Defaultability
+
+* Any numeric value type is defaultable
+
+* A reference value type is defaultable if it is not of the form `ref $t`
+
+* Locals must have a type that is defaultable.
+
+* Table definitions with non-zero minimum size must have an element type that is defaultable. (Imports are not affected.)
+
+
+### Instructions
+
+#### Equality
+
+* `ref.eq` compares two references whose types support equality
+  - `ref.eq : [eqref eqref] -> [i32]`
+
+
+#### Functions
+
+* `ref.func` creates a function reference from a function index
+  - `ref.func $f : [] -> (ref $t)`
+     - iff `$f : $t`
+
+* `call_ref` calls a function through a reference
+  - `call_ref : [t1* (optref $t)] -> [t2*]`
+     - iff `$t = [t1*] -> [t2*]`
+  - traps on `null`
+
+
+#### Structures
+
+* `struct.new <typeidx>` allocates a structure of type `$t` and initialises its fields with given values
+  - `struct.new $t : [t*] -> [(ref $t)]`
+    - iff `$t = struct (mut t)*`
+
+* `struct.new_default <typeidx>` allocates a structure of type `$t` and initialises its fields with default values
+  - `struct.new_default $t : [] -> [(ref $t)]`
+    - iff `$t = struct (mut t)*`
+    - and all `t*` are defaultable
+
+* `struct.get <typeidx> <fieldidx>` reads field `$x` from a structure
+  - `struct.get $t i : [(optref $t)] -> [t]`
+    - iff `$t = struct (mut1 t1)^i (mut ti) (mut2 t2)*`
+    - and `t = unpacked(ti)`
+  - traps on `null`
```

Comment:
This confuses me: the static type of the referred-to object appears to be restricted to $t exactly, but there is no widening / upcast operator that I can see.  Are you expecting widening casts to be handled by LET in some fashion or by changes to type compatibility for eg SET_LOCAL?  Or is it just a matter of the missing prose in the section on Value Conversions?

I'd also be curious to hear about why you feel <typeidx> is a desirable operand here, since the verifier will know the static type of the operand.

Reply:
This follows the usual declarative style of having a separate subsumption rule. For the Wasm type system it is built into the sequencing rule (see the spec draft for the reference types proposal for details; concretely, the last rule in [this subsection](https://webassembly.github.io/reference-types/core/valid/instructions.html#instruction-sequences)). With this approach, individual rules can always "assume" the correct type.

Your second question seems to be missing a word? Did you mean the type immediate? A stated design criterion for Wasm was to always have all operationally relevant type information explicit in the instructions (except where the semantics is completely parametric in a type), e.g. no overloading. So this is just following that.

Reply:
OK, that's fine.  So the subtyping rules a little further up here really extend the `match rule there.

(For the second para I did indeed mean the type immediate; github helpfully removed a word in angle brackets from my comment.)